server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -147,149 +147,208 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
-        final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
-
-        boolean hasIndexRequestsWithPipelines = false;
-        final MetaData metaData = clusterService.state().getMetaData();
-        ImmutableOpenMap<String, IndexMetaData> indicesMetaData = metaData.indices();
-        for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
-            IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
-            if (indexRequest != null) {
-                // get pipeline from request
-                String pipeline = indexRequest.getPipeline();
-                if (pipeline == null) {
-                    // start to look for default pipeline via settings found in the index meta data
-                    IndexMetaData indexMetaData = indicesMetaData.get(actionRequest.index());
-                    // check the alias for the index request (this is how normal index requests are modeled)
-                    if (indexMetaData == null && indexRequest.index() != null) {
-                        AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
-                        if (indexOrAlias != null && indexOrAlias.isAlias()) {
-                            AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
-                            indexMetaData = alias.getWriteIndex();
-                        }
+        new BulkExecutor(task, bulkRequest, listener).run();
+    }
+
+    /**
+     * A runnable that will ensure the cluster state has been recovered enough to
+     * read index metadata and templates/pipelines.  Will retry up to the bulk's timeout
+     */
+    private final class BulkExecutor extends ActionRunnable<BulkResponse> {
+        private final ClusterStateObserver recoveredObserver;
+        private final BulkRequest bulkRequest;
+        private final Task task;
+        long startTime = -1;
+
+        BulkExecutor(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+            super(listener);
+            this.recoveredObserver = new ClusterStateObserver(clusterService, bulkRequest.timeout(), logger, threadPool.getThreadContext());
+            this.bulkRequest = bulkRequest;
+            this.task = task;
+        }
+
+        @Override
+        protected void doRun() {
+            if (startTime == -1) {
+               startTime = relativeTime();
+            }
+            ClusterBlockException blockException = clusterService.state().blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+            if (blockException != null) {
+                if (recoveredObserver.isTimedOut()) {
+                    // we running as a last attempt after a timeout has happened. don't retry
+                    listener.onFailure(blockException);
+                    return;
+                }
+                recoveredObserver.waitForNextChange(new ClusterStateObserver.Listener() {
```

Comment:
I think we could use `waitForNextChange(Listener listener, Predicate<ClusterState> statePredicate)` to filter out any intermediate states that still have a block - that way we don't need to check that again and can go straight into `prepForBulk`